Sources/Workspace/ManagedDependency.swift

```diff
@@ -14,195 +14,160 @@ import SourceControl
+        public func edited(unmanagedPath: AbsolutePath?) -> ManagedDependency {
+            return .edited(packageRef: self.packageRef, basedOn: self, unmanagedPath: unmanagedPath)
+        }
 
-    internal init(
-        packageRef: PackageReference,
-        state: State,
-        subpath: RelativePath,
-        basedOn: ManagedDependency?
-    ) {
-        self.packageRef = packageRef
-        self.subpath = subpath
-        self.basedOn = basedOn
-        self.state = state
-    }
+        /// Create a dependency present locally on the filesystem.
+        public static func local(
+            packageRef: PackageReference
+        ) -> ManagedDependency {
+            return ManagedDependency(
+                packageRef: packageRef,
+                state: .local
+            )
+        }
 
-    private init(
-        basedOn dependency: ManagedDependency,
-        subpath: RelativePath,
-        unmanagedPath: AbsolutePath?
-    ) {
-        assert(dependency.state.isCheckout)
-        self.basedOn = dependency
-        self.packageRef = dependency.packageRef
-        self.subpath = subpath
-        self.state = .edited(unmanagedPath)
-    }
+        /// Create a remote dependency checked out
+        public static func remote(
+            packageRef: PackageReference,
+            state: CheckoutState
+        ) -> ManagedDependency {
+            return ManagedDependency(
+                packageRef: packageRef,
+                state: .checkout(state)
+            )
+        }
 
-    /// Create an editable managed dependency based on a dependency which
-    /// was *not* in edit state.
-    ///
-    /// - Parameters:
-    ///     - subpath: The subpath inside the editable directory.
-    ///     - unmanagedPath: A custom absolute path instead of the subpath.
-    public func editedDependency(subpath: RelativePath, unmanagedPath: AbsolutePath?) -> ManagedDependency {
-        return ManagedDependency(basedOn: self, subpath: subpath, unmanagedPath: unmanagedPath)
+        /// Create an edited dependency
+        public static func edited(
+            packageRef: PackageReference,
+            basedOn: ManagedDependency?,
+            unmanagedPath: AbsolutePath?
+        ) -> ManagedDependency {
+            return ManagedDependency(
+                packageRef: packageRef,
+                state: .edited(basedOn: basedOn, unmanagedPath: unmanagedPath)
+            )
+        }
     }
 }
 
-extension ManagedDependency: CustomStringConvertible {
+extension Workspace.ManagedDependency: CustomStringConvertible {
     public var description: String {
         return "<ManagedDependency: \(self.packageRef.name) \(self.state)>"
     }
 }
 
-
 // MARK: - ManagedDependencies
 
-/// A collection of managed dependencies.
-public final class ManagedDependencies {
-    /// The dependencies keyed by the package URL.
-    private var dependencyMap: [String: ManagedDependency]
```

Comment:
@mattt this is where I got in trouble with my refactoring. to make incremental progress, I suggest we keep the map based on location right now (ie [String: ManagedDependency]) and then I will find a way to refactor / fix the manifest loading logic in Workspace so we can finally change this to [PackageIdentity: ManagedDependency]

wdyt?

Reply:
@tomerd That works for me. To help keep our history comprehensible, could you patch #3698 to isolate the call site changes and force push? (Maybe you have a local branch that with unsquashed commits for adb61b3?)

Did #3749 resolved any of your trouble on manifest loading? We might be able to keep keying `dependencies` on `PackageIdentity` since it's equivalent to what we were doing before, comparing the `lowercased()` versions, like in `subscript(forNameOrIdentity:)`

Reply:
this is now done, I went for a bit of a different approach. #3749 helps in various other places, but he the case is more subtle with how manifests are loaded (long story)